learner/sampler/slurper.py

```diff
@@ -0,0 +1,134 @@
+        if not encoding:
+            encoding = self.encoding            
+
+        try:
+            with open(file, encoding = encoding) as f:
+                f.seek(1000,0)
+                f.readline()
+                f.close()
+            return encoding
+        except:
+            common_encodings = ["Latin-1", "UTF-16", "ascii", "cp037", "cp437", "UTF-32"]
+
+            for codec in common_encodings:
+                try:
+                    with open(file, encoding = codec) as f:
+                        f.readline()
+                        f.close()
+                    return codec
+                except:
+                    continue
+            print("Your file is of unusual type - can you specify the encoding for us?")
+    
+    def estimate_line_size(file: str = None) -> int:
+        """Estimate the number of lines in the file- requires exactly 62 line reads."""
+        if not file:
+            file = self.filename
+            
+        with open(file) as f:
+            #chop header
+            f.readline()
+            line = f.readline()
+            line_length = len(line)
+            f.seek(0,2)
+            eof = f.tell()
+            average_length = 0
+            for sample in range(2,32):
+                f.seek(math.floor(nlines/sample),0)
+                f.readline()
+                line = f.readline()
+                average_length += len(line)
+            f.close()
+        return math.floor(eof/(average_length/30))
+            
+    def get_sample_size(nlines: int = None, confidence: float = 0.95, error: float = 0.05) -> int:
+        """Returns number of samples needed for a population of nlines, when sampling is done with replacement"""
+        if not nlines:
+            nlines = self.nlines
+            
+        #Hard coding the critical and std. dev. for now
+        X = (1.96 * 0.5 * 0.5)/(error**2)
+        return math.floor(nlines*X / (X + N - 1))
+            
+    
+    def pythonic_reservoir(file: str = None, reservoir_size: int = None) -> List[str]:
+        """Make a single pass through the file, replacing each value with some probability. This is Knuth's Reservoir Sampling."""
+        if not file:
+            file = self.filename
+        if not reservoir_size:
+            reservoir_size = self.sample_size
+            
+        reservoir = []
+        n = reservoir_size - 1
+        counter = 0
+        
+        with open(file) as f:
+            for line in f:
+                if counter < reservoir_size:
+                    reservoir.append(line)
+                else:
+                    n += 1
+                    draw = random.randrange(0,n,1)
+                    if draw < reservoir_size:
+                        reservoir[draw] = line
+            f.close()
+        
+        return reservoir
+                    
+                    
+    def pythonic_read_random_lines(file: str = None, sample_size: int = None, byte_bite: int = 20) -> List[str]:
```

Comment:
Cool stuff! I am going to take it for granted for now that the reservoir sampling actually works. Will need to see further testing tomorrow.